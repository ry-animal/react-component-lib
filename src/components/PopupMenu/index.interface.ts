import { ReactElement, ReactNode, RefObject } from "react";
import { DefaultTheme } from "styled-components";
import { CircleButtonProps, TextButtonProps } from "../Buttons";

export interface ListItem {
  onClick: () => void;
  icon?: ReactElement | string;
  label: string;
}

export enum ALIGN {
  left = "left",
  right = "right",
}

export interface PopupMenuStrings {
  buttonLabel: string;
  closeButtonLabel: string;
}

export interface MobileMenuProps {
  /**
   * Action to handle close event
   */
  onClose: () => void;
  /**
   * Menu options
   */
  children: ReactNode[];
  /**
   * String constants
   */
  strings: PopupMenuStrings;
  /**
   * Theme because outside of theme hierarchy
   */
  theme: DefaultTheme;
}

export interface DesktopMenuProps {
  alignTo: ALIGN;
  menuOptions: JSX.Element[];
  onClose: () => void;
  buttonRef: RefObject<HTMLElement>;
}

export interface PopupMenuProps {
  /**
   * Align menu to 'left' or 'right' of button
   */
  alignTo?: ALIGN;
  /**
   * Show close icon when menu open
   */
  disableCloseIcon?: boolean;
  /**
   * Button props
   */
  button: CircleButtonProps | TextButtonProps;
  /**
   * Menu options
   */
  options: ListItem[];
  /**
   * String constants
   */
  strings?: PopupMenuStrings;
  /**
   * Overrides existing icon for close, takes an url to icon
   */
  closeIconUrl?: string;
}
