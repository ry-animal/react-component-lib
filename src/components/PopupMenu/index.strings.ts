export default {
  buttonLabel: "Open Menu",
  closeButtonLabel: "Close Menu",
};
