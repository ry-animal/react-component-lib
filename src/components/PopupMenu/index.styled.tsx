import { rgba } from "polished";
import styled from "styled-components";
import { BlurRadius } from "../../constants/BlurRadius";
import Icon from "../Icon";
import { ALIGN } from "./index.interface";

export const MenuButtonWrapper = styled.div`
  position: relative;
`;

export const TextWrapper = styled.span`
  ${({ theme }) => `
    padding: ${theme.space.s4};
    color: ${theme.color.TEXT_1};
  `}
`;

export const OptionsWrapper = styled.div<{ alignTo: ALIGN }>`
  width: 200px;
  box-sizing: border-box;
  max-height: 180px;
  overflow-y: auto;
  position: absolute;
  display: flex;
  flex-direction: column;
  align-items: flex-start;

  ${({ theme }) => `
    padding: ${theme.space.s8};
    border-radius: ${theme.radius.r8};
    margin-top: ${theme.space.s4};
    gap: ${theme.space.s4};
    z-index: ${theme.layer.tooltip};
    background-color: ${rgba(theme.color.SURFACE_2, theme.opacity.NINE)};
    box-shadow: ${theme.elevation.medium};
  `}

  ${({ alignTo }) => {
    if (alignTo === ALIGN.right) {
      return "right: 0;";
    } else {
      return "left: 0;";
    }
  }}
`;

export const OptionButton = styled.button<{ backgroundColor?: string }>`
  border: none;
  width: 100%;
  cursor: pointer;
  display: flex;
  align-items: center;
  text-align: left;
  transition: background-color 0.25s ease-out;

  ${({ backgroundColor, theme }) => `
    background-color: ${backgroundColor ?? theme.color.SURFACE_2};
    padding: ${theme.space.s4};
    border-radius: ${theme.radius.r6};
    font-family: ${theme.type.family.body};
  `}

  @media (max-width: ${({ theme }) => theme.media.mobile.max}) {
    font-size: ${({ theme }) => theme.type.size.SM};
  }

  &:disabled {
    background: transparent;
    opacity: 0.6;
    cursor: default;

    border: ${({ theme }) =>
      `${theme.space.s1} solid ${theme.color.SURFACE_3};`};
  }

  &:hover,
  &:focus {
    background-color: ${({ theme }) => theme.color.SURFACE_3};
    @supports (backdrop-filter: blur(${BlurRadius.b15})) {
      backdrop-filter: blur(${BlurRadius.b15});
    }
  }
`;

export const CustomIcon = styled(Icon)`
  max-width: 22px;
  margin-right: ${({ theme }) => theme.space.s8};
`;

export const MobileMenuWrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  align-items: flex-end;

  ${({ theme }) => `
    background-color: ${rgba(theme.color.BLACK, theme.opacity.FIVE)};
    z-index: ${theme.layer.popover};
  `}
`;

export const CloseButtonWrapper = styled.div`
  position: absolute;

  ${({ theme }) => `
    right: ${theme.space.s16};
    top: ${theme.space.s16};
  `}
`;

export const PopupMenuWrapper = styled.div`
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  flex: 1;
  display: flex;
  flex-direction: column;
  max-height: 50vh;
  overflow-y: auto;
  ${({ theme }) => `
    background: ${theme.color.SURFACE_2};
    border-top-left-radius: ${theme.radius.r16};
    border-top-right-radius: ${theme.radius.r16};
    padding: ${theme.space.s32};
    gap: ${theme.space.s16};
  `}
`;
