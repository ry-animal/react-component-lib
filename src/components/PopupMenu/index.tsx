import { ReactElement, useEffect, useRef, useState } from "react";
import ReactDOM from "react-dom";
import { ThemeProvider, useTheme } from "styled-components";
import CloseIconWhite from "../../assets/icons/icon-clear-white.svg";
import CloseIconBlack from "../../assets/icons/icon-clear.svg";
import { useClosePopup } from "../../hooks/useClosePopup";
import CircleButton from "../Buttons/CircleButton";
import { TextButton } from "../Buttons/TextButton";
import ClickOutside from "../ClickOutside";
import DisplayOnMedia from "../DisplayOnMedia";
import { IconClear } from "../Icons";
import IconMoreOptions from "../Icons/IconMoreOptions";
import {
  ALIGN,
  DesktopMenuProps,
  MobileMenuProps,
  PopupMenuProps,
} from "./index.interface";
import STRINGS from "./index.strings";
import * as Styled from "./index.styled";

const MobileMenu = ({ onClose, children, strings, theme }: MobileMenuProps) => (
  <ThemeProvider theme={theme}>
    <Styled.MobileMenuWrapper data-testid="popup-mobile-menu">
      <ClickOutside onClick={onClose}>
        <Styled.CloseButtonWrapper>
          <CircleButton
            icon={<IconClear />}
            onClick={onClose}
            aria-label={strings.closeButtonLabel}
          />
        </Styled.CloseButtonWrapper>
        <Styled.PopupMenuWrapper>{children}</Styled.PopupMenuWrapper>
      </ClickOutside>
    </Styled.MobileMenuWrapper>
  </ThemeProvider>
);

const DesktopMenu = ({
  alignTo,
  menuOptions,
  onClose,
  buttonRef,
}: DesktopMenuProps) => {
  const desktopRef = useRef<HTMLDivElement>(null);
  useClosePopup([desktopRef, buttonRef], onClose);
  return (
    <Styled.OptionsWrapper
      ref={desktopRef}
      alignTo={alignTo}
      data-testid="popup-menu"
    >
      {menuOptions}
    </Styled.OptionsWrapper>
  );
};

const attemptRemoveMobileMenu = () => {
  const modal = document.getElementById("popup-mobile-menu");
  modal?.remove();
};

const getCloseIcon = (overrideIcon?: string, buttonDarkTheme?: boolean) => {
  if (overrideIcon) {
    return overrideIcon;
  }
  return buttonDarkTheme ? CloseIconWhite : CloseIconBlack;
};

export const PopupMenu = ({
  button,
  options,
  disableCloseIcon,
  alignTo = ALIGN.left,
  strings = STRINGS,
  closeIconUrl: closeIconOverride,
}: PopupMenuProps) => {
  const theme = useTheme();
  const [isMenuOpen, setMenuOpen] = useState(false);
  const buttonRef = useRef<HTMLButtonElement>(null);

  useEffect(() => {
    attemptRemoveMobileMenu();

    return () => attemptRemoveMobileMenu();
  }, []);

  const closeMenu = () => {
    attemptRemoveMobileMenu();
    setMenuOpen(false);
  };

  const onOptionSelected = (callback: () => void) => {
    callback();
    closeMenu();
  };

  const closeIconUrl = getCloseIcon(closeIconOverride, button.isDarkTheme);

  const determineIconUrl = (buttonIcon?: string | ReactElement) => {
    if (isMenuOpen && !disableCloseIcon) {
      return closeIconUrl;
    }
    return buttonIcon;
  };

  const menuOptions = options.map((item, index) => (
    <Styled.OptionButton
      key={index}
      onClick={() => onOptionSelected(item.onClick)}
      aria-label={item.label}
    >
      {item.icon && (
        <Styled.CustomIcon icon={item.icon} alt={`${item.label} icon`} />
      )}
      <Styled.TextWrapper>{item.label}</Styled.TextWrapper>
    </Styled.OptionButton>
  ));

  function createMobileMenu() {
    attemptRemoveMobileMenu();

    const modal = document.createElement("div");
    modal.setAttribute("id", "popup-mobile-menu");

    ReactDOM.render(
      <DisplayOnMedia showOn="mobile">
        <MobileMenu onClose={closeMenu} strings={strings} theme={theme}>
          {menuOptions}
        </MobileMenu>
      </DisplayOnMedia>,
      document.body.appendChild(modal)
    );
  }

  useEffect(() => {
    if (isMenuOpen) {
      createMobileMenu();
    } else {
      attemptRemoveMobileMenu();
    }
  }, [isMenuOpen]);

  const toggleMenuOpen = () => setMenuOpen((prevState) => !prevState);
  return (
    <Styled.MenuButtonWrapper>
      {"label" in button ? (
        <TextButton
          {...button}
          icon={determineIconUrl(button.icon ?? <IconMoreOptions />)}
          aria-label={button["aria-label"] ?? strings.buttonLabel}
          onClick={toggleMenuOpen}
          forwardRef={buttonRef}
        />
      ) : (
        <CircleButton
          {...button}
          iconUrl={determineIconUrl(button.iconUrl ?? button.icon)}
          aria-label={button["aria-label"] ?? strings.buttonLabel}
          onClick={toggleMenuOpen}
          forwardRef={buttonRef}
        />
      )}

      {isMenuOpen && (
        <>
          <DisplayOnMedia hideOn="mobile">
            <DesktopMenu
              alignTo={alignTo}
              menuOptions={menuOptions}
              onClose={closeMenu}
              buttonRef={buttonRef}
            />
          </DisplayOnMedia>
        </>
      )}
    </Styled.MenuButtonWrapper>
  );
};
export default PopupMenu;
