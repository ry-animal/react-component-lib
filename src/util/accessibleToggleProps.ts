export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
