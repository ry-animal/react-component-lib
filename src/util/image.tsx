import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
