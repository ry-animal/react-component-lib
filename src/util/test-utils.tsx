import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
