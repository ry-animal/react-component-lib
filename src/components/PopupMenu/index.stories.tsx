import { ComponentMeta, ComponentStory } from "@storybook/react";
import { BrowserRouter } from "react-router-dom";
import styled from "styled-components";
import PopupMenu from ".";
import EthIcon from "../../assets/icons/icon-chain-eth.svg";
import LrcIcon from "../../assets/icons/icon-chain-lrc.svg";
import ChevronDown from "../../assets/icons/icon-chevron-down.svg";
import MoreOptionsIcon from "../../assets/icons/icon-more-option.svg";
import { ALIGN } from "./index.interface";

export default {
  component: PopupMenu,
  title: "Components/PopupMenu",
} as ComponentMeta<typeof PopupMenu>;

const BackgroundContrast = styled.div`
  width: 100%;
  height: 70vh;
`;

const defaultArgs = {
  options: [
    {
      onClick: () => {},
      label: "Option 1",
      icon: EthIcon,
    },
    {
      onClick: () => {},
      label: "Option 2",
      icon: LrcIcon,
    },
  ],
  button: {
    iconUrl: MoreOptionsIcon,
    transparent: true,
  },
};

const Template: ComponentStory<typeof PopupMenu> = (args) => (
  <BackgroundContrast>
    <BrowserRouter>
      <PopupMenu {...args} />
    </BrowserRouter>
  </BackgroundContrast>
);

export const DefaultPopupMenu = Template.bind({});
DefaultPopupMenu.args = {
  ...defaultArgs,
};

export const TextButtonWithCloseIconOverride = Template.bind({});
TextButtonWithCloseIconOverride.args = {
  ...defaultArgs,
  button: {
    label: "Wallet",
    icon: ChevronDown,
    isDarkTheme: true,
    size: "medium",
    flipIcon: true,
  },
  closeIconUrl: ChevronDown,
};

export const DisableCloseIcon = Template.bind({});
DisableCloseIcon.args = {
  ...defaultArgs,
  disableCloseIcon: true,
};

export const ManyItems = Template.bind({});
ManyItems.args = {
  ...defaultArgs,
  options: [
    ...defaultArgs.options,
    ...defaultArgs.options,
    ...defaultArgs.options,
    ...defaultArgs.options,
  ],
};

const RightAlignedTemplate: ComponentStory<typeof PopupMenu> = (args) => (
  <BackgroundContrast>
    <BrowserRouter>
      <div
        style={{ width: "100%", display: "flex", justifyContent: "flex-end" }}
      >
        <PopupMenu {...args} />
      </div>
    </BrowserRouter>
  </BackgroundContrast>
);

export const RightAligned = RightAlignedTemplate.bind({});
RightAligned.args = {
  ...defaultArgs,
  alignTo: ALIGN.right,
};

export const TextButtonAlignRight = RightAlignedTemplate.bind({});
TextButtonAlignRight.args = {
  ...RightAligned.args,
  ...TextButtonWithCloseIconOverride.args,
};
