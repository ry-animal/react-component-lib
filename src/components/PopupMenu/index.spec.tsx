import userEvent from "@testing-library/user-event";
import PopupMenu from ".";
import ChevronDown from "../../assets/icons/icon-chevron-down.svg";
import { fireEvent, render, screen, waitFor } from "../../util/test-utils";
import { ALIGN, PopupMenuProps } from "./index.interface";
import STRINGS from "./index.strings";

describe("<PopupMenu />", () => {
  const mockMedia = {
    media: "",
    onchange: null,
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
  };

  const sampleProps = (): PopupMenuProps => ({
    options: [
      { onClick: jest.fn(), icon: "", label: "Manage listings" },
      { onClick: jest.fn(), icon: "", label: "Show NFT" },
    ],
    button: {
      iconUrl: "./icon-url.svg",
      transparent: true,
      onClick: jest.fn(),
    },
  });

  describe("rendering", () => {
    test("Default rendering matches snapshot", () => {
      expect(
        render(<PopupMenu {...sampleProps()} />).asFragment()
      ).toMatchSnapshot();
    });

    test("Rendering matches snapshot (right aligned)", () => {
      expect(
        render(
          <PopupMenu {...sampleProps()} alignTo={ALIGN.right} />
        ).asFragment()
      ).toMatchSnapshot();
    });

    test("Rendering matches snapshot (TextButton)", () => {
      expect(
        render(
          <PopupMenu
            {...sampleProps()}
            button={{
              label: "Text",
              onClick: jest.fn(),
              "aria-label": "Text Button",
            }}
          />
        ).asFragment()
      ).toMatchSnapshot();
    });

    test("Rendering matches snapshot (Circle Button)", () => {
      expect(
        render(
          <PopupMenu
            {...sampleProps()}
            button={{
              icon: "./icon-url.svg",
              isDarkTheme: true,
              onClick: jest.fn(),
              "aria-label": "Circle Button",
            }}
          />
        ).asFragment()
      ).toMatchSnapshot();
    });

    test("Rendering matches snapshot (disableCloseIcon)", () => {
      expect(
        render(
          <PopupMenu {...sampleProps()} disableCloseIcon={true} />
        ).asFragment()
      ).toMatchSnapshot();
    });

    test("overrides close icon", () => {
      const button = render(
        <PopupMenu
          {...sampleProps()}
          closeIconUrl={ChevronDown}
          button={{
            label: "Text",
            onClick: jest.fn(),
            "aria-label": "Text Button",
          }}
        />
      ).getByLabelText("Text Button");

      expect(button).not.toBeNull();
    });
  });

  describe("event handling", () => {
    test("Clicking the More Options icon opens menu", async () => {
      window.matchMedia = jest.fn().mockImplementation((query) => ({
        ...mockMedia,
        matches: query === "(min-width: 1025px) and (max-width: 1279px)",
      }));

      render(<PopupMenu {...sampleProps()} />);

      fireEvent.click(screen.getByLabelText(STRINGS.buttonLabel));

      await waitFor(() => screen.getByTestId("popup-menu"));
      expect(screen.getByText("Manage listings")).toBeInTheDocument();
      expect(screen.getByText("Show NFT")).toBeInTheDocument();
    });

    test("Clicking More Options icon opens menu and option is clickable", async () => {
      window.matchMedia = jest.fn().mockImplementation((query) => ({
        ...mockMedia,
        matches: query === "(min-width: 1025px) and (max-width: 1279px)",
      }));
      const onClick = jest.fn();
      render(
        <PopupMenu
          {...sampleProps()}
          options={[
            { onClick, icon: "", label: "Manage listings" },
            { onClick: jest.fn(), icon: "", label: "Show NFT" },
          ]}
        />
      );
      fireEvent.click(screen.getByLabelText(STRINGS.buttonLabel));

      await waitFor(() => screen.getByTestId("popup-menu"));
      expect(screen.getByText("Manage listings")).toBeInTheDocument();

      fireEvent.click(screen.getByText("Manage listings"));
      expect(onClick).toBeCalled();
    });

    test("Clicking More Options icon opens MOBILE menu and option is clickable", async () => {
      window.matchMedia = jest.fn().mockImplementation((query) => ({
        ...mockMedia,
        matches: query === "(max-width: 767px)",
      }));
      const onClick = jest.fn();
      render(
        <PopupMenu
          {...sampleProps()}
          options={[
            { onClick, icon: "", label: "Manage listings" },
            { onClick: jest.fn(), icon: "", label: "Show NFT" },
          ]}
        />
      );
      fireEvent.click(screen.getByLabelText(STRINGS.buttonLabel));

      await waitFor(() => screen.getByTestId("popup-mobile-menu"));
      expect(screen.getByText("Manage listings")).toBeInTheDocument();

      fireEvent.click(screen.getByText("Manage listings"));
      expect(onClick).toBeCalled();
    });

    test("Mobile popup menu is closeable", async () => {
      window.matchMedia = jest.fn().mockImplementation((query) => ({
        ...mockMedia,
        matches: query === "(max-width: 767px)",
      }));
      const customStrings = {
        buttonLabel: "Button Label",
        closeButtonLabel: "Close",
      };
      render(<PopupMenu {...sampleProps()} strings={customStrings} />);
      fireEvent.click(screen.getByLabelText(customStrings.buttonLabel));

      expect(screen.queryByTestId("popup-mobile-menu")).toBeInTheDocument();

      fireEvent.click(screen.getByLabelText(customStrings.closeButtonLabel));
      expect(screen.queryByTestId("popup-mobile-menu")).toBeNull();
    });

    test("TextButton is clickable and opens menu", async () => {
      window.matchMedia = jest.fn().mockImplementation((query) => ({
        ...mockMedia,
        matches: query === "(max-width: 767px)",
      }));
      render(
        <PopupMenu
          {...sampleProps()}
          button={{ label: "Text", onClick: jest.fn() }}
        />
      );
      fireEvent.click(screen.getByLabelText(STRINGS.buttonLabel));

      expect(screen.queryByTestId("popup-mobile-menu")).toBeInTheDocument();

      fireEvent.click(screen.getByLabelText(STRINGS.buttonLabel));

      expect(screen.queryByTestId("popup-mobile-menu")).toBeNull();
    });

    test("Default align adds left style", async () => {
      window.matchMedia = jest.fn().mockImplementation((query) => ({
        ...mockMedia,
        matches: query === "(min-width: 1025px) and (max-width: 1279px)",
      }));
      render(<PopupMenu {...sampleProps()} />);
      fireEvent.click(screen.getByLabelText(STRINGS.buttonLabel));

      expect(screen.queryByTestId("popup-menu")).toHaveStyleRule("left", "0");
    });

    test("Align Right adds right style", async () => {
      window.matchMedia = jest.fn().mockImplementation((query) => ({
        ...mockMedia,
        matches: query === "(min-width: 1025px) and (max-width: 1279px)",
      }));
      render(<PopupMenu {...sampleProps()} alignTo={ALIGN.right} />);
      fireEvent.click(screen.getByLabelText(STRINGS.buttonLabel));

      expect(screen.queryByTestId("popup-menu")).toHaveStyleRule("right", "0");
    });

    test("useClosePopup working for large screens", async () => {
      window.matchMedia = jest.fn().mockImplementation((query) => ({
        ...mockMedia,
        matches: query === "(min-width: 1025px) and (max-width: 1279px)",
      }));
      render(<PopupMenu {...sampleProps()} alignTo={ALIGN.right} />);
      fireEvent.click(screen.getByLabelText(STRINGS.buttonLabel));

      expect(screen.queryByTestId("popup-menu")).toBeInTheDocument();

      userEvent.click(document.body);

      expect(screen.queryByTestId("popup-menu")).toBeNull();
    });
  });
});
