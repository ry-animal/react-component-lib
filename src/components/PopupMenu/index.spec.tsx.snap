// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PopupMenu /> rendering Default rendering matches snapshot 1`] = `
<DocumentFragment>
  .c1 {
  cursor: pointer;
  border: none;
  border-radius: 10000px;
  box-shadow: 0px 2px 8px RGBA(0,0,0,0.04);
  pointer-events: auto;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FFFFFF;
  width: 40px;
  height: 40px;
  background-size: 20px auto;
  background-image: url(./icon-url.svg);
  background-repeat: no-repeat;
  background-position: center;
  background-color: rgba(255,255,255,0.7);
  -webkit-backdrop-filter: blur(15px);
  backdrop-filter: blur(15px);
}

.c1:disabled {
  pointer-events: none;
  cursor: default;
}

.c1:hover {
  opacity: 100%;
  background-color: rgba(255,255,255,0.9);
}

.c0 {
  position: relative;
}

<div
    class="c0"
  >
    <button
      aria-label="Open Menu"
      class="c1"
    />
  </div>
</DocumentFragment>
`;

exports[`<PopupMenu /> rendering Rendering matches snapshot (Circle Button) 1`] = `
<DocumentFragment>
  .c1 {
  cursor: pointer;
  border: none;
  border-radius: 10000px;
  box-shadow: 0px 2px 8px RGBA(0,0,0,0.04);
  pointer-events: auto;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FFFFFF;
  width: 40px;
  height: 40px;
  background-size: 20px auto;
  background-image: url(./icon-url.svg);
  background-repeat: no-repeat;
  background-position: center;
}

.c1:disabled {
  pointer-events: none;
  cursor: default;
}

.c1:hover {
  opacity: 100%;
}

.c2 {
  width: 80%;
}

.c0 {
  position: relative;
}

<div
    class="c0"
  >
    <button
      aria-label="Circle Button"
      class="c1"
    >
      <img
        aria-label="Icon Button icon"
        class="c2"
        src="./icon-url.svg"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`<PopupMenu /> rendering Rendering matches snapshot (TextButton) 1`] = `
<DocumentFragment>
  .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 svg {
  width: 100%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: unset;
  -ms-flex: unset;
  flex: unset;
  line-height: 1;
  gap: 20px;
}

.c1 {
  text-align: center;
  border: none;
  outline: none;
  overflow: hidden;
  cursor: pointer;
  pointer-events: auto;
  -webkit-transition: background-color 0.25s ease-out;
  transition: background-color 0.25s ease-out;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-height: 72px;
  font-weight: 600;
  padding: 0 28px;
  font-size: 18px;
  border-radius: 4px;
  background-color: #000000;
  color: #EEEEEE;
}

.c1:disabled {
  opacity: 20%;
  cursor: default;
  pointer-events: none;
}

.c1:hover:not([disabled]),
.c1:focus-visible:not([disabled]) {
  background-color: #5F5F5F;
}

.c0 {
  position: relative;
}

<div
    class="c0"
  >
    <button
      aria-label="Text Button"
      class="c1"
    >
      <span
        class="c2"
      >
        <div
          alt="Text icon"
          class="c3"
        >
          <svg
            fill="white"
            height="4"
            viewBox="0 0 18 4"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 3C9.55228 3 10 2.55228 10 2C10 1.44772 9.55228 1 9 1C8.44772 1 8 1.44772 8 2C8 2.55228 8.44772 3 9 3Z"
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M16 3C16.5523 3 17 2.55228 17 2C17 1.44772 16.5523 1 16 1C15.4477 1 15 1.44772 15 2C15 2.55228 15.4477 3 16 3Z"
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M2 3C2.55228 3 3 2.55228 3 2C3 1.44772 2.55228 1 2 1C1.44772 1 1 1.44772 1 2C1 2.55228 1.44772 3 2 3Z"
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        Text
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<PopupMenu /> rendering Rendering matches snapshot (disableCloseIcon) 1`] = `
<DocumentFragment>
  .c1 {
  cursor: pointer;
  border: none;
  border-radius: 10000px;
  box-shadow: 0px 2px 8px RGBA(0,0,0,0.04);
  pointer-events: auto;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FFFFFF;
  width: 40px;
  height: 40px;
  background-size: 20px auto;
  background-image: url(./icon-url.svg);
  background-repeat: no-repeat;
  background-position: center;
  background-color: rgba(255,255,255,0.7);
  -webkit-backdrop-filter: blur(15px);
  backdrop-filter: blur(15px);
}

.c1:disabled {
  pointer-events: none;
  cursor: default;
}

.c1:hover {
  opacity: 100%;
  background-color: rgba(255,255,255,0.9);
}

.c0 {
  position: relative;
}

<div
    class="c0"
  >
    <button
      aria-label="Open Menu"
      class="c1"
    />
  </div>
</DocumentFragment>
`;

exports[`<PopupMenu /> rendering Rendering matches snapshot (right aligned) 1`] = `
<DocumentFragment>
  .c1 {
  cursor: pointer;
  border: none;
  border-radius: 10000px;
  box-shadow: 0px 2px 8px RGBA(0,0,0,0.04);
  pointer-events: auto;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #FFFFFF;
  width: 40px;
  height: 40px;
  background-size: 20px auto;
  background-image: url(./icon-url.svg);
  background-repeat: no-repeat;
  background-position: center;
  background-color: rgba(255,255,255,0.7);
  -webkit-backdrop-filter: blur(15px);
  backdrop-filter: blur(15px);
}

.c1:disabled {
  pointer-events: none;
  cursor: default;
}

.c1:hover {
  opacity: 100%;
  background-color: rgba(255,255,255,0.9);
}

.c0 {
  position: relative;
}

<div
    class="c0"
  >
    <button
      aria-label="Open Menu"
      class="c1"
    />
  </div>
</DocumentFragment>
`;
